src/hotspot/share/cds/aotConstantPoolResolver.cpp

```diff
@@ -318,6 +325,201 @@ void AOTConstantPoolResolver::maybe_resolve_fmi_ref(InstanceKlass* ik, Method* m
+          // We initialize the class of the archived lambda proxy at VM start-up, which will also initialize
+          // the interface that it implements. If that interface has a clinit method, we can potentially
+          // change program execution order.
+          if (log_is_enabled(Debug, cds, resolve)) {
+            ResourceMark rm;
+            log_debug(cds, resolve)("Cannot aot-resolve Lambda proxy of interface type %s (has <cilint>)", k->external_name());
+          }
+          return false;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+bool AOTConstantPoolResolver::check_lambda_metafactory_methodtype_arg(ConstantPool* cp, int bsms_attribute_index, int arg_i) {
+  int mt_index = cp->operand_argument_index_at(bsms_attribute_index, arg_i);
+  if (!cp->tag_at(mt_index).is_method_type()) {
+    // malformed class?
+    return false;
+  }
+
+  Symbol* sig = cp->method_type_signature_at(mt_index);
+  if (log_is_enabled(Debug, cds, resolve)) {
+    ResourceMark rm;
+    log_debug(cds, resolve)("Checking MethodType for LambdaMetafactory BSM arg %d: %s", arg_i, sig->as_C_string());
+  }
+
+  return check_lambda_metafactory_signature(cp, sig, false);
+}
+
+bool AOTConstantPoolResolver::check_lambda_metafactory_methodhandle_arg(ConstantPool* cp, int bsms_attribute_index, int arg_i) {
+  int mh_index = cp->operand_argument_index_at(bsms_attribute_index, arg_i);
+  if (!cp->tag_at(mh_index).is_method_handle()) {
+    // malformed class?
+    return false;
+  }
+
+ Symbol* sig = cp->method_handle_signature_ref_at(mh_index);
+  if (log_is_enabled(Debug, cds, resolve)) {
+    ResourceMark rm;
+    log_debug(cds, resolve)("Checking MethodType of MethodHandle for  LambdaMetafactory BSM arg %d: %s", arg_i, sig->as_C_string());
+  }
+  if (!check_lambda_metafactory_signature(cp, sig, false)) {
+    return false;
+  }
+
+  Symbol* mh_sig = cp->method_handle_signature_ref_at(mh_index);
+  if (!check_lambda_metafactory_signature(cp, mh_sig, false)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool AOTConstantPoolResolver::is_indy_resolution_deterministic(ConstantPool* cp, int cp_index) {
+  assert(cp->tag_at(cp_index).is_invoke_dynamic(), "sanity");
+  if (!CDSConfig::is_dumping_invokedynamic()) {
+    return false;
+  }
+
+  InstanceKlass* pool_holder = cp->pool_holder();
+  if (!SystemDictionaryShared::is_builtin(pool_holder)) {
+    return false;
+  }
+
+  int bsm = cp->bootstrap_method_ref_index_at(cp_index);
+  int bsm_ref = cp->method_handle_index_at(bsm);
+  Symbol* bsm_name = cp->uncached_name_ref_at(bsm_ref);
+  Symbol* bsm_signature = cp->uncached_signature_ref_at(bsm_ref);
+  Symbol* bsm_klass = cp->klass_name_at(cp->uncached_klass_ref_index_at(bsm_ref));
+
+  // We currently support only StringConcatFactory::makeConcatWithConstants() and LambdaMetafactory::metafactory()
+
+  if (bsm_klass->equals("java/lang/invoke/StringConcatFactory") &&
+      bsm_name->equals("makeConcatWithConstants")) {
+    return true;
```

Comment:
I think all BSM entries in SCF are deterministic, really. So we can just check for the class?

Reply:
We haven't done much testing with the other SCF BSMs, so I think it's better to do that in a follow-up REF.

Reply:
`StringConcatFactory​::makeConcat(…)` is definitely deterministic, as it simply delegates to `StringConcatFactory​::makeConcatWithConstants(…)` with the recipe being based on the number of `MethodType` parameters:
https://github.com/openjdk/jdk/blob/de90204b60c408ef258a2d2515ad252de4b23536/src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java#L231-L240

And there is no other bootstrap method in `StringConcatFactory` at the moment.

Reply:
Since the comment says "This bootstrap method is unlikely to be used in practice", I am not sure if we can sufficiently test it (with aot-linking).